spec.bs

```diff
@@ -243,8 +258,189 @@ Each {{URLPattern}} object has an associated <dfn for=URLPattern>hash component<
+
+  1. Let |result| be the result of [=match=] given [=this=], |input|, and |baseURL| if given.
+  1. If |result| is null, return false.
+  1. Return true.
+</div>
+
+<div algorithm>
+  The <dfn method for="URLPattern">exec(|input|, |baseURL|)</dfn> method steps are:
+
+  1. Return the result of [=match=] given [=this=], |input|, and |baseURL| if given.
+</div>
+
+<div algorithm>
+  To perform a <dfn>match</dfn> given a {{URLPattern}} |urlpattern|, a {{URLPatternInput}} |input|, and an optional string |baseURLString|:
+
+  1. Let |protocol| be the empty string.
+  1. Let |username| be the empty string.
+  1. Let |password| be the empty string.
+  1. Let |hostname| be the empty string.
+  1. Let |port| be the empty string.
+  1. Let |pathname| be the empty string.
+  1. Let |search| be the empty string.
+  1. Let |hash| be the empty string.
+  1. Let |inputs| be an empty [=list=].
+  1. [=list/Append=] |input| on to |inputs|.
+  1. If |input| is a {{URLPatternInit}} then:
+    1. If |baseURLString| was given, throw a {{TypeError}}.
+    1. Let |applyResult| be the result of [=process a URLPatternInit=] given |input|, "url", |protocol|, |username|, |password|, |hostname|, |port|, |pathname|, |search|, and |hash|. If this throws an exception, catch it, and return null.
+    1. Set |protocol| to |applyResult|["{{URLPatternInit/protocol}}"].
+    1. Set |username| to |applyResult|["{{URLPatternInit/username}}"].
+    1. Set |password| to |applyResult|["{{URLPatternInit/password}}"].
+    1. Set |hostname| to |applyResult|["{{URLPatternInit/hostname}}"].
+    1. Set |port| to |applyResult|["{{URLPatternInit/port}}"].
+    1. Set |pathname| to |applyResult|["{{URLPatternInit/pathname}}"].
+    1. Set |search| to |applyResult|["{{URLPatternInit/search}}"].
+    1. Set |hash| to |applyResult|["{{URLPatternInit/hash}}"].
+  1. Else:
+    1. Let |baseURL| be null.
+    1. If |baseURLString| was given, then:
+      1. Set |baseURL| to the result of [=URL parser|parsing=] |baseURLString|.
+      1. If |baseURL| is failure, return null.
+      1. Push |baseURL| on to |inputs|.
+    1. Let |url| be the result of [=URL parser|parsing=] |input| given |baseURL|.
+    1. If |url| is failure, return null.
+    1. Set |protocol| to |url|'s [=url/scheme=].
+    1. Set |username| to |url|'s [=url/username=].
+    1. Set |password| to |url|'s [=url/password=].
+    1. Set |hostname| to |url|'s [=url/host=] or the empty string if the value is null.
+    1. Set |port| to |url|'s [=url/port=] or the empty string if the value is null.
+    1. Set |pathname| to |url|'s [=url/API pathname string=].
+    1. Set |search| to |url|'s [=url/query=] or the empty string if the value is null.
+    1. Set |hash| to |url|'s [=url/fragment=] or the empty string if the value is null.
+  1. Let |protocolExecResult| be the result of [$RegExpBuiltinExec$] given |urlpattern|'s [=URLPattern/protocol component=]'s [=component/regular expression=] and |protocol|.
```

Comment:
You could optionally write these in function call form, i.e.
```suggestion
  1. Let |protocolExecResult| be [$RegExpBuiltinExec$](|urlpattern|'s [=URLPattern/protocol component=]'s [=component/regular expression=], |protocol|).
```

Up to you which you find clearer.